extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+# workspace/reference extension to LSP
+
+The `workspace/reference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references that code inside of the workspace root makes to code symbols that are external to the workspace root (e.g. dependencies).
+
+Use case: clients of a language server can invoke `workspace/reference` in order to find external references to e.g. imported dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate client-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.uri` which is located within the workspace.
+- Excluding any `ReferenceInformation.name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
```

Comment:
> it would not be feasible to do a separate request for each reference to get the SymbolInformation, right?

correct

- I've changed `location` -> `reference`. I think this makes more sense than `referenceLocation` because you can write e.g. `refs[0].reference.uri` or `refs[0].reference.range` instead of more verbose form `.referenceLocation.range` etc.
- Using a full `SymbolInformation` is tricky because we don't want to force a language server to fully resolve the dependency. Check out the latest revision of the doc I push for an update on this.